beeline-core/src/main/java/io/honeycomb/beeline/tracing/README.md

```diff
@@ -0,0 +1,144 @@
+final Response response = mapTarget.request(MediaType.APPLICATION_JSON_TYPE)
+    .header("x-honeycomb-trace", MessageFormat.format("1;trace_id={0},parent_id={1}", span.getTraceId(), span.getParentSpanId()))
+    .get();
+```
+
+To do this programmatically, implement ClientRequestFilter / ClientResponseFilter
+
+### Java HTTP client
+Java SDK added a built-in HTTP client. It was experimental in JDK 9 and became stable in JDK 11.
+
+#### Synchronous
+Example creating request with Honeycomb Trace set.
+```java
+HttpClient client = HttpClient.newBuilder().build();
+HttpRequest request = HttpRequest.newBuilder()
+    .uri(new URI("https://postman-echo.com/get"))
+    .header("x-honeycomb-trace", MessageFormat.format("1;trace_id={0},parent_id={1}", span.getTraceId(), span.getParentSpanId()))
+    .GET()
+    .build();
+HttpResponse<String> response = client.send(request, HttpResponse.BodyHandler.asString())
+// process response
+```
+#### Asynchronous
+Example creating request with Honeycomb Trace set.
+```java
+HttpClient client = HttpClient.newBuilder().build();
+HttpRequest request = HttpRequest.newBuilder()
+    .uri(new URI("https://postman-echo.com/get"))
+    .header("x-honeycomb-trace", MessageFormat.format("1;trace_id={0},parent_id={1}", span.getTraceId(), span.getParentSpanId()))
+    .GET()
+    .build();
+CompletableFuture<HttpResponse<String>> future = client.sendAsync(request, HttpResponse.BodyHandler.asString());
+HttpResponse<String> response = future.get();
+// process response
+```
+### Resteasy
+Resteasy supports both JAX-RS (Jersey) and Apache HttpClient implementation clients. Syntax is based on the JAX-RS specification.
+
+Example creating request with Honeycomb Trace header set using default client (Jersey).
+```java
+Client client = ResteasyClientBuilder.newBuilder().build();
+WebTarget target = client.target("http://google.com");
+Invocation.Builder request = target.request();
+request.header("x-honeycomb-trace", MessageFormat.format("1;trace_id={0},parent_id={1}", span.getTraceId(), span.getParentSpanId()))
+Response response = request.get();
+```
+### Dropwizard
+Dropwizard's HTTP client is a Jersey client that also uses Apache HttpClient. As a Jersey client, syntax is based on the JAX-RS specification.
+
+Example creating request with Honeycomb Trace header set.
+```java
+// `environment` is an instance of io.dropwizard.setup.Environment
+// jerseyConfiguration is an instance of io.dropwizard.client.JerseyClientConfiguration
+JerseyClientBuilder builder = new io.dropwizard.client.JerseyClientBuilder(environment);
+client = builder.using(jerseyConfiguration).build("example");
+// Once client is created, use JAX-RS API as normal
+final WebTarget target = client.target("https://google.com").path("/maps");
+final Response response = target.request(MediaType.APPLICATION_JSON_TYPE)
+    .header("x-honeycomb-trace", MessageFormat.format("1;trace_id={0},parent_id={1}", span.getTraceId(), span.getParentSpanId()))
+    .get();
```

Comment:
Love the examples, and the range of clients, but I'm wondering if we can simplify using available API constants methods to generate the header.

Reply:
Updated to use constant instead of hardcoded `"x-honeycomb-trace"`

Reply:
Better, but still requires the user to know the internal format of the wire protocol (including the version, etc). It also doesn't show off all the features of the header. I know it's more code, but I was thinking of something like:

```
HttpHeaderV1PropagationCodec codec = HttpHeaderV1PropagationCodec.getInstance();
Map<String, Object> context = Stream.of(new Object[][] {
            { "user_id", 1123 },
            { "user_name", "alyssaphacker" },
}).collect(Collectors.toMap(data -> (String) data[0], data -> data[1]));
Optional<String> headerValue = codec.encode(
    new PropagationContext(span.getTraceId(), span.getParentSpanId(), "dataset", context));
final Response response = target.request(MediaType.APPLICATION_JSON_TYPE)
    .header(HttpHeaderV1PropagationCodec.HONEYCOMB_TRACE_HEADER, headerValue.orElse(""))
    ....

```

Reply:
You could simplify, or make it more idiomatic, but it demonstrates:

* Sending dataset, context fields
* Using encode to safely convert a PropagationContext into a header value

Without requiring the caller to know the format of our trace header.

Reply:
I have  pulled the implementation portion out of each example and put it into a Reference utility function. I believe this enhances the conciseness and readability.